### ISSUE TITLE
cmd/7g: bad codegen, probably regopt related

### ISSUE THREAD

**Issue by mwhudson** [golang/go#10253]
Distilled from one of the failures in go test strings, a small-ish test case looks like this:

```shellsession
mwhudson@am2:~/go/src$ cat s5.go
package main

import "fmt"

type stringFinder struct {
	pattern        string
}

var f = &stringFinder{pattern: "He"}

func next() int {
	i := 1
	for i < 6 {
		// Compare backwards from the end until the first unmatching character.
		j := 1
		for j >= 0 && "xHello"[i] == f.pattern[j] {
			i--
			j--
		}
		if j < 0 {
			return i + 1 // match
		}
		i += 1
	}
	return -1
}

func main() {
	fmt.Println(next())
}
mwhudson@am2:~/go/src$ ~/go/bin/go run s5.go
-1
mwhudson@am2:~/go/src$ ~/go1.4/bin/go run s5.go
1
```

This assembles to https://gist.github.com/mwhudson/41bbb7f10f6eda55ad86, which is as hard to read as any page of assembly, but single stepping through in gdb shows the problem: line 25 loads "xHellox"[i] into w0 and line 32 stomps on it, before line 39 does the comparison.

**Comment by 4ad** [comment]
This looks as if RtoB is not used in some place, but rather some manual RtoB lookalike.

**Comment by 4ad** [comment]
Well, the problem is that register variables are allocated from R0 up instead of R7, and they collide with the temp registers; this can only happen if RtoB and friends are not consistently used, however, that's a different bug. The real issue here is what causes register variables to be allocated from R0.

**Comment by 4ad** [comment]
It looks like paint2 does not realise that R2 should not be used in a
particular region. I don't understand how that is possible.

R2 is used in the previous region, and usedreg correctly indicates so,
but in the next region it can't find R2.

**Comment by 4ad** [comment]
Simpler reproduction:


```
package main

func main() {    
        text := "abc"
        s := &s{text}
        println(text[2] == s.text[2])
}

type s struct {
        text string
}
```

**Comment by minux** [comment]
is the problem that cmd/gc uses R0, R1, ...
instead of only using register starting from arm64.REGMIN?

**Comment by 4ad** [comment]
No. Starting from R7 just masks the problem away, but it should work with R0 as well.

**Comment by crawshaw** [comment]
In an attempt to get some readable assembly, this program (which still incorrectly prints false):

```
package main

var text = "abc"
var s = &str{text}

func main() {
        println(eq())
}

//go:nosplit
func eq() bool {
        return text[0] == s.text[0]
}

type str struct {
        text string
}
```

Produces with ```go run -gcflags "-l -B -S" junk.go```:

```
".eq t=1 nosplit size=112 value=0 args=0x8 locals=0x0 leaf
        0x0000 00000 (/home/crawshaw/junk.go:11)        TEXT    "".eq(SB), $-8-8
        0x0000 00000 (/home/crawshaw/junk.go:11)        FUNCDATA        ZR, gclocals·23e8278e2b69a3a75fa59b23c49ed6ad(SB)
        0x0000 00000 (/home/crawshaw/junk.go:11)        FUNCDATA        $1, gclocals·33cdeccccebe80329f1fdbee7f5874cb(SB)
        0x0000 00000 (/home/crawshaw/junk.go:11)        MOVD    ZR, R0
        0x0004 00004 (/home/crawshaw/junk.go:11)        MOVBU   R0, R0
        0x0008 00008 (/home/crawshaw/junk.go:12)        MOVD    "".text(SB), R0
        0x0010 00016 (/home/crawshaw/junk.go:12)        MOVBU   (R0), R0
        0x0014 00020 (/home/crawshaw/junk.go:12)        MOVD    "".s(SB), R1
        0x001c 00028 (/home/crawshaw/junk.go:12)        CMP     R1, ZR
        0x0020 00032 (/home/crawshaw/junk.go:12)        BNE     40
        0x0024 00036 (/home/crawshaw/junk.go:12)        MOVD    ZR, (R1)
        0x0028 00040 (/home/crawshaw/junk.go:12)        MOVD    (R1), R2
        0x002c 00044 (/home/crawshaw/junk.go:12)        MOVD    R2, R0
        0x0030 00048 (/home/crawshaw/junk.go:12)        MOVD    8(R1), R2
        0x0034 00052 (/home/crawshaw/junk.go:12)        MOVD    R0, R1
        0x0038 00056 (/home/crawshaw/junk.go:12)        MOVBU   (R1), R1
        0x003c 00060 (/home/crawshaw/junk.go:12)        CMP     R1, R0
        0x0040 00064 (/home/crawshaw/junk.go:12)        BEQ     80
        0x0044 00068 (/home/crawshaw/junk.go:12)        MOVD    ZR, R0
        0x0048 00072 (/home/crawshaw/junk.go:12)        MOVBU   R0, "".~r0(FP)
        0x004c 00076 (/home/crawshaw/junk.go:12)        RET     (R30)
        0x0050 00080 (/home/crawshaw/junk.go:12)        MOVD    $1, R0
        0x0054 00084 (/home/crawshaw/junk.go:12)        MOVBU   R0, "".~r0(FP)
        0x0058 00088 (/home/crawshaw/junk.go:12)        RET     (R30)
        0x005c 00092 (/home/crawshaw/junk.go:12)        JMP     0(PC)
        0x0060 00096 (/home/crawshaw/junk.go:12)        DWORD   "".text(SB)
        0x0068 00104 (/home/crawshaw/junk.go:12)        DWORD   "".s(SB)
```

**Comment by 4ad** [comment]
It appears that paint2, at some point, returns the wrong register set. It returns only what's used inside the current region. It does not account for the fact that it previously registered something into the same register already.

**Comment by crawshaw** [comment]
golang.org/cl/8321

**Comment by crawshaw** [comment]
Checking out commit ```11dba2ec2d8f``` (last known ok build) I notice that the generated unoptimized code is studiously avoiding using R0 anywhere. It also has a tendency to start functions at R7. After the breaking change, register numbering becomes far more normal, beginning at R0.

I modified the implementation of cgen's OINDEX so it looks like it did at ```11dba2ec2d8f```. The result was the exact same code for the implementation of ```eq```, except the registers are numbered from R0 instead of R7. The resulting programs crash very early in the runtime trying to index into arrays.

My suspicion is the object code generated for R0 is wrong. But it seems neither 7l -a nor objdump work with arm64?? How are you debugging output?

**Comment by 4ad** [comment]
The R7 vs. R0 thing is a red herring. Traditionally, the temp set was slightly larger than the reg var set. Perhaps this mattered at some point, but it didn't matter now, so @rsc removed this distinction when consolidating all registerizers into the portable registerizer, making the two sets equal.

This exposed this bug. The bug was (probably) present before, but because there was a big enough difference between the two sets, temporary registers never clashed with register variables. Now they compete for the same registers and the bug is visible.

7l -a hasn't worked in over a year and a half since we have had liblink or its Go successor. Objdump, GNU objdump that is, works with arm64 binaries provided you compile it with support for the target. Obviously the native objdump present on arm64 linux has this property, but you can compile binutils with arm64 support on any system.

**Comment by 4ad** [comment]
Found the problem, fix soon.

**Comment by crawshaw** [comment]
I didn't realize you were still working on this.

Looks like we came up with an answer at the same time, I just mailed https://golang.org/cl/8387

**Comment by crawshaw** [comment]
A bit more detail to go with that CL:

In the program I posted above the bug comes long before paint2. The wrong value of rgp.varno is being fed in, showing R0 as unused when in fact it is needed after the region being optimized for the CMP.

A look at the debugging output for pass1 shows:

```0:00043 (/Users/crawshaw/junk.go:12)    CMP     R1, R0   u1:.R1(1)```

The problem is R0 should be on the use1 list, but it's not. This use of the second (I saw it somewhere referred to as "middle" register) is caught on other architectures by this logic in mkvar:

```
                // Compute used register for reg
                if p.Info.Flags&RegRead != 0 {
                        r.use1.b[0] |= Thearch.RtoB(int(p.Reg))
                }
```

So I propose setting the RegRead flag for CMP.

**Comment by 4ad** [comment]
ARM has the same problem, but we don't see it because the peep optimizer remove the loads that would otherwise affect the comparisons.

We need a test that fails on arm as well and put it in test/fixedbugs.

**Comment by crawshaw** [comment]
Thanks @4ad, much appreciated. I think the arm fix should be a separate CL, both for simplicity and because I don't have an arm machine handy to test a fix on. If you have one, do you want to take it?

Also this doesn't quite fix the build, as there is still #10304.

**Comment by 4ad** [comment]
I don't have an arm machine handy. Arm64 can run arm binaries, though, provided you have a kernel with 4k pages.

I use a 64k kernel on my arm64 because I want to make sure Go works with 64k pages. I can revert it to a 4k kernel for this, but this week and the next one I am very very busy, so if someone else can do it it would be great.
